Implement allocator context-locality

eliaskoromilas opened the pull request:
- [x] Transform `current_handler` to a `PyContextVar` which stores a `PyCapsule` of the current `PyDataMem_Handler`.
- [x] Initialize the `current_handler` with a  `PyCapsule` of the `default_handler`.
- [x] Change `PyDataMem_GetHandlerName` to `PyDataMem_GetHandler` to avoid a new API call that would just return the current `PyDataMem_Handler`.
- [x] Add tests that demonstrate allocator context-locality (also thread-locality).
- [x] Fix documentation

**Notes**:
* During `PyDataMem_SetHandler` a capsule is created to hold the handler, with the context var keeping the only reference to it.
* A subsequent `PyDataMem_SetHandler` call releases the old capsule (GC) and replaces it with the new one.

Also includes the following `test_mem_policy` enhancements:
* Rename `set_new_policy` to `set_secret_data_policy` and make it return a capsule of the old policy (instead of its name only)
* Modify `set_old_policy` to accept a capsule of the policy to set, `None` to set to the "default_allocator" policy
* Purge `test_prefix` since it's covered by `np.core.multiarray.get_handler_name()`

eliaskoromilas commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -3093,6 +3093,12 @@ VOID_compare(char *ip1, char *ip2, PyArrayObject *ap)
     if (!PyArray_HASFIELDS(ap)) {
         return STRING_compare(ip1, ip2, ap);
     }
+    const PyDataMem_Handler *mem_handler = PyDataMem_GetHandler(NULL);
+    if (mem_handler == NULL) {
+        /* Better fallback to default than goto finish. */
+        PyErr_Clear();
+        mem_handler = &default_handler;
+    }
```
`VOID_compare` returns `int`. On error, the function finishes prematurely. Should this happen here too (returning res=0)? I didn't find any reference to it, where it's being used?

eric-wieser replied:
Yes, I think returning early with an error set is the right thing to do.

eric-wieser replied:
You can't find a reference to it because `VOID_compare` is likely used in templated code as something like `@type@_compare` or `VOID_@op@`.

eric-wieser commented on numpy/core/src/multiarray/alloc.c:
```diff
@@ -470,50 +470,80 @@ PyDataMem_UserRENEW(void *ptr, size_t size, const PyDataMemAllocator *allocator)
+    PyDataMem_Handler *handler;
     if (obj == NULL) {
-        return current_handler->name;
+        if (PyContextVar_Get(current_handler, NULL, &capsule)) {
```
This can set `capsule=NULL` based on my reading of the docs, in which case this should probably error.

eliaskoromilas replied:
In this case, the context var has a `default_value`, so it will always return something.
https://github.com/eliaskoromilas/numpy/blob/configurable_allocator/numpy/core/src/multiarray/multiarraymodule.c#L4918

eric-wieser replied:
Another concern; does `PyContextVar_New` require the GIL to be held?

eliaskoromilas replied:
Not to my knowledge.

https://github.com/python/cpython/blob/main/Python/context.c#L806-L842

eliaskoromilas commented:
@mattip I had to merge my `mem_policy` test changes with your https://github.com/numpy/numpy/pull/17582/commits/a7a543556807d199913c86584121ac09860353aa commit.

I didn't see that on time. I had already fixed those signatures but I also had renamed the `Allocator` struct to `SecretDataAllocatorFuncs` to make it more self-explanatory. I added some comments, too.

mattip commented:
Thanks. There is apparently a problem with [cygwin build](https://github.com/numpy/numpy/pull/17582/checks?check_run_id=3278842337) on the original PR, I think the extbuild tests need to add a link library for python. I am still looking.

eliaskoromilas commented:
@mattip I think I merged them just fine.

Regarding the cygwin build, I have no clue but I could also take a look at it.

If you rebase `main` won't I be able to run here the same checks? If yes, is there a problem doing it? Thanks.

mattip commented on doc/source/reference/c-api/data_memory.rst:
```diff
@@ -79,19 +79,17 @@ reallocate or free the data memory of the instance.
-   that will be used to allocate data for the next ``PyArrayObject``.
+   Return the `PyDataMem_Handler` used by the ``PyArrayObject``. If ``NULL``,
+   return the current global policy that will be used to allocate data for the
+   next ``PyArrayObject``. On failure, set an exception and return ``NULL``.
```
```suggestion
   next ``PyArrayObject``. On failure, return ``NULL``.
```

eliaskoromilas replied:
~Most of~ the involved calls in these 2 functions (e.g. `PyCapsule_New`) set an exception (a `PyErr`) on error.

Your suggestion is not to document that our functions`set an exception`, or also to clear any exception (`PyErr_Clear`) before returning NULL?

mattip replied:
When I look at the implementation of `PyDataMem_GetHandler`, I do not see any code that is setting an exception, so that is not something this function is doing. Whether an exception is set or not is the realm of  `PyContextVar_Get`, `PyCapsule_GetPointer`, and `PyArray_HANDLER`.

eliaskoromilas replied:
What you are saying is true. Indeed claiming that on failure an exception is set is incorrect, since there are cases that this doesn't happen.

On the other hand, if our functions used functions that always `set an exception` on failure, I think that the natural way would be to document it.

I agree that my description is misleading, I will commit your suggestions. Thanks.

mattip commented on numpy/core/src/multiarray/alloc.c:
```diff
@@ -470,50 +470,80 @@ PyDataMem_UserRENEW(void *ptr, size_t size, const PyDataMemAllocator *allocator)
- * the user-provided functions so they will still call the python
- * and numpy memory management callback hooks.
+ * Set a new allocation policy. If the input value is NULL, will reset
+ * the policy to the default. Return the previous policy, or set an exception
+ * and return NULL if an error has occurred. We wrap the user-provided
```
```suggestion
 * the policy to the default. Return the previous policy, or
 * return NULL if an error has occurred. We wrap the user-provided
```

eliaskoromilas replied:
`PyContextVar_Get/Set` don't set an exception on failure
`PyCapsule_GetPointer/New` set an exception on failure

mattip commented on numpy/core/src/multiarray/alloc.c:
```diff
@@ -470,50 +470,80 @@ PyDataMem_UserRENEW(void *ptr, size_t size, const PyDataMemAllocator *allocator)
- * will be used to allocate data for the next PyArrayObject.
+ * Return the PyDataMem_Handler used by the PyArrayObject. If NULL, return
+ * the current global policy that will be used to allocate data
+ * for the next PyArrayObject. On failure, set an exception and return NULL.
```
```suggestion
 * for the next PyArrayObject. On failure, return NULL.
```

eliaskoromilas replied:
`PyContextVar_Get` doesn't set an exception on failure
`PyCapsule_GetPointer` sets an exception on failure

mattip commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2153,7 +2157,10 @@ array_setstate(PyArrayObject *self, PyObject *args)
         }
         /* Store the functions in case the global handler is modified */
-        fa->mem_handler = current_handler;
+        fa->mem_handler = (PyDataMem_Handler *) PyDataMem_GetHandler(NULL);
```
We need the cast because of the `const`? Maybe it would be better to declare `mem_handler` with a `const` qualifier instead?

eliaskoromilas replied:
True. I think we should better `const` the struct field.

eliaskoromilas replied:
On a second look, since we `tp_alloc` the `PyArrayObject_fields` struct, making the field `const` would just require a double cast to set it.

```
*(PyDataMem_Handler *)&fa->mem_handler = (PyDataMem_Handler *) PyDataMem_GetHandler(NULL);
```
I believe it's best to keep it as is. What do you think?

mattip commented:
Windows has some newly unmatched alloc/free pairs. I see [`uh-oh, unmatched shift_free(ptr, 8) but allocated 720`](https://dev.azure.com/mattipicus/mattipicus/_build/results?buildId=106&view=logs&j=e630afe2-84ca-5052-3b94-af602773dcac&t=1ab64c83-2999-50d7-9b3d-2ee9888a7657&l=7400) at the end of testing, and some test failures.

mattip commented:
Ahh, the same unmatched pair message [shows up on linux](https://dev.azure.com/mattipicus/mattipicus/_build/results?buildId=106&view=logs&j=c4fbb33d-b115-53c9-54fa-705790f154ff&t=1437d177-db37-57b8-1516-a6c596dd8a61&l=6865)

mattip commented:
Grr, the same failures are on the original PR as well.

eliaskoromilas commented:
Yes, there are always 3 unmatched pairs. I think I have also brought it up in a discussion.

eliaskoromilas commented:
I noticed the Windows test failures but I wasn't sure if they are related to this PR.

Can we sync the test suite with the upstream? The `master` in this repository is way behind `upstream:main`.

eliaskoromilas commented on doc/source/reference/c-api/data_memory.rst:
```diff
@@ -79,19 +79,17 @@ reallocate or free the data memory of the instance.
-   the policy to the default. Returns the previous policy, ``NULL`` if the
-   previous policy was the default. We wrap the user-provided functions
+   Set a new allocation policy. If the input value is ``NULL``, will reset the
+   policy to the default. Return the previous policy, or set an exception and
```
```suggestion
   policy to the default. Return the previous policy, or
```

eliaskoromilas commented:
[Windows Python39-64bit-full > Run NumPy Test Suite](https://dev.azure.com/mattipicus/mattipicus/_build/results?buildId=108&view=logs&j=e630afe2-84ca-5052-3b94-af602773dcac&t=1ab64c83-2999-50d7-9b3d-2ee9888a7657) got fixed by itself ???

mattip commented:
Thanks @eliaskoromilas
